### PR TITLE
fix: timeoutRunner with large monotonicTime

### LINKED ISSUE #24432: [internal] fix internal debian 11/12 ARM tests
https://github.com/microsoft/playwright/commit/879ddb73b063e9ff97654b4e3d92969f84f201c8#diff-9fe1fe455371e5592c4ac04ef88de1ae4331ce749bd6333afe8734bb4f7d8b4e

### PR DESCRIPTION
**Pull request by mxschmitt**
The issue was that the monotonicTime on some of our servers was `4832315541.968` while the `MaxTime` was `2147483647`.  This created a negative timeout for the TimeoutRunner and caused the timeout exceeded error.

Fixes https://github.com/microsoft/playwright/issues/24432.

### CODE REVIEW

File: packages/playwright-test/src/worker/testInfo.ts [hunk 1]
```diff
@@ -109,7 +109,7 @@ export class TestInfoImpl implements TestInfo {
 
   _deadlineForMatcher(timeout: number): { deadline: number, timeoutMessage: string } {
     const startTime = monotonicTime();
-    const matcherDeadline = timeout ? startTime + timeout : MaxTime;
+    const matcherDeadline = timeout ? startTime + timeout : MaxSetIntervalValue;
```
Comment (pavelfeldman) [inline packages/playwright-test/src/worker/testInfo.ts:112]: Deadline can't be capped with MaxSetIntervalValue if monotonicTime can be bigger than MaxSetIntervalValue.

File: packages/playwright-test/src/worker/timeoutManager.ts [hunk 2]
```diff
@@ -166,4 +166,11 @@ export class TimeoutManager {
       stack: location ? message + `\n    at ${location.file}:${location.line}:${location.column}` : undefined
     };
   }
+
+  static calculateMinimalTimeout(...timeouts: number[]): number {
```
Comment (dgozman) [inline packages/playwright-test/src/worker/timeoutManager.ts:170]:
```suggestion
  static calculateEarliestDeadline(...deadlines: number[]): number {
```

File: packages/playwright-test/src/worker/testInfo.ts [hunk 3]
```diff
@@ -109,11 +109,19 @@ export class TestInfoImpl implements TestInfo {
...
   _deadlineForMatcher(timeout: number): { deadline: number, timeoutMessage: string } {
     const startTime = monotonicTime();
-    const matcherDeadline = timeout ? startTime + timeout : MaxTime;
+    const matcherDeadline = timeout ? startTime + timeout : 0;
     const testDeadline = this._timeoutManager.currentSlotDeadline() - 100;
     const matcherMessage = `Timeout ${timeout}ms exceeded while waiting on the predicate`;
     const testMessage = `Test timeout of ${this.timeout}ms exceeded`;
-    return { deadline: Math.min(testDeadline, matcherDeadline), timeoutMessage: testDeadline < matcherDeadline ? testMessage : matcherMessage };
+    const deadline = this._calculateEarliestDeadline(matcherDeadline, testDeadline);
+    return { deadline, timeoutMessage: deadline === matcherDeadline ? matcherMessage : testMessage };
```
Comment (dgozman) [inline packages/playwright-test/src/worker/testInfo.ts:117]: Now there is just a small chance of the wrong error message - when `matcherDeadline === testDeadline`.

File: packages/playwright-test/src/worker/testInfo.ts [hunk 4]
```diff
@@ -109,11 +109,19 @@ export class TestInfoImpl implements TestInfo {
 
   _deadlineForMatcher(timeout: number): { deadline: number, timeoutMessage: string } {
     const startTime = monotonicTime();
-    const matcherDeadline = timeout ? startTime + timeout : MaxTime;
+    const matcherDeadline = timeout ? startTime + timeout : 0;
     const testDeadline = this._timeoutManager.currentSlotDeadline() - 100;
```
Comment (pavelfeldman) [inline packages/playwright-test/src/worker/testInfo.ts:113]: currentSlotDeadline needs to be fixed too as it is using MaxTime.

**Comment by pavelfeldman** [comment]
I think it would be easier to use Number.MAX_SAFE_INTEGER for MaxTime and cap the actual setTimeout calls with 2^31

**Comment by pavelfeldman** [comment]
Introducing a public `setDeadline(deadline, callback, now?: number)` in timeoutRunner would also help managing all the deadline-based timeouts.

**Comment by mxschmitt** [comment]
Superseded by https://github.com/microsoft/playwright/pull/24518

**Pull request closed without merging**